docs/source/app-dev/command-deduplication.rst

```diff
@@ -0,0 +1,364 @@
+
+
+Failure scenarios
+-----------------
+
+The above strategy can fail in the following scenarios:
+
+#. The bound ``B`` is too low: The command can be executed multiple times.
+   
+   Possible causes:
+
+   - You have retried for longer than the deduplication duration, but never got a meaningful answer, e.g., because the timeout (gRPC deadline) is too short.
+     For example, this can happen due to long-running Daml interpretation when using the :ref:`Command Service <command-service>`.
+
+   - The application clock drifts significantly from the participant's or ledger's clock.
+
+   - There are unexpected network delays.
+
+   - Submissions are retried internally in the participant or Daml ledger and those retries do not stop before ``B`` is over.
+     Refer to the specific ledger's documentation for more information.
+
+#. Unacceptable changes cause infinite retries
+
+   You need business workflow knowledge to decide that retrying does not make sense any more.
+   Of course, you can always stop retrying and accept that you do not know the outcome for sure.
+
+
+Unknown processing time bounds
+==============================
+
+Finding a good bound ``B`` on the processing time is hard, and there may still be unforeseen circumstances that delay processing beyond the chosen bound ``B``.
+You can avoid these problems by using deduplication offsets instead of durations.
+An offset defines a point in the history of the ledger and is thus not affected by clock skews and network delays.
+Offsets are arguably less intuitive and require more effort by the application developer.
+We recommend the following strategy for using deduplication offsets:
+
+#. Choose a fresh command ID for the ledger change and the ``actAs`` parties, which (together with the application ID) determine the change ID.
+   Remember the command ID across application crashes.
+   (Analogous to :ref:`Step 1 above <dedup-bounded-step-command-id>`)
+
+   .. _dedup-unbounded-step-dedup-offset:
+   
+#. Obtain a recent offset ``OFF0`` on the completion event stream and remember across crashes that you use ``OFF0`` with the chosen command ID. There are several ways to do so:
+
+   - Use the :ref:`Command Service <command-service>` to obtain a recent offset by repeatedly submitting a dummy command, e.g., a :ref:`Create-And-Exercise command <com.daml.ledger.api.v1.CreateAndExerciseCommand>` of some single-signatory template with the :ref:`Archive <function-da-internal-template-functions-archive-52202>` choice, until you get a successful response.
+     The response contains the :ref:`completion offset <com.daml.ledger.api.v1.SubmitAndWaitForTransactionIdResponse.completion_offset>`.
+
+   - Use the :ref:`Command Completion Service <command-completion-service>` by asking for the :ref:`current ledger end <com.daml.ledger.api.v1.CommandCompletionService.CompletionEnd>`.
```

Comment:
From a practical point of view, the latter point is simpler to implement so I'd suggest reversing the order of these options for obtaining a recent offset.

Reply:
Reversed.